# _Encryptix Internship_: __Data Science__
## __Task 1:__ Classify the Genre of the  movie based on it's Description 
### Technique: TF-IDF with Naive Bayes
1. Import Libraries & Load the Dataset
2. Preprocessing using Spacy
3. Train Test Split
4. Training the model using Naive Bayes approach
5. Evaluating the model/ Generating Classification report

### __STEP 1__: _Load the Dataset_

In [1]:
#importing libraries
import pandas as pd
from sklearn.metrics import classification_report
from sklearn.model_selection import train_test_split


In [2]:
df = pd.read_excel('Moviesdata.xlsx')
df.head()

Name       Genre  \
0       Oscar et la dame rose (2009)       drama    
1                       Cupid (1997)    thriller    
2   Young, Wild and Wonderful (1980)       adult    
3              The Secret Sin (1915)       drama    
4             The Unrecovered (2007)       drama    

                                         Description Unnamed: 3 Unnamed: 4  \
0   Listening in to a conversation between his do...        NaN        NaN   
1   A brother and sister with a past incestuous r...        NaN        NaN   
2   As the bus empties the students for their fie...        NaN        NaN   
3   To help their unemployed father make ends mee...        NaN        NaN   
4   The film's title refers not only to the un-re...        NaN        NaN   

  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ... Unnamed: 29  \
0        NaN        NaN        NaN        NaN        NaN  ...         NaN   
1        NaN        NaN        NaN        NaN        NaN  ...         NaN   
2        NaN        NaN        NaN        NaN        NaN  ...         NaN   
3        NaN        NaN        NaN        NaN        NaN  ...         NaN   
4        NaN        NaN        NaN        NaN        NaN  ...         NaN   

  Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34 Unnamed: 35  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2         NaN         NaN         NaN         NaN         NaN         NaN   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 36 Unnamed: 37 Unnamed: 38  
0         NaN         NaN         NaN  
1         NaN         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN  

[5 rows x 39 columns]

In [ ]:
df.info()

In [3]:
df.drop(df.columns[3:39], axis=1, inplace=True)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48079 entries, 0 to 48078
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         24029 non-null  object
 1   Genre        21454 non-null  object
 2   Description  21000 non-null  object
dtypes: object(3)
memory usage: 1.1+ MB


In [5]:
# Check for missing values in each column
missing_values = df.isnull().sum()
print(missing_values)


Name           24050
Genre          26625
Description    27079
dtype: int64


In [6]:
df = df.dropna()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20999 entries, 0 to 48078
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Name         20999 non-null  object
 1   Genre        20999 non-null  object
 2   Description  20999 non-null  object
dtypes: object(3)
memory usage: 656.2+ KB


In [8]:
genre_mapping = {
    'action': 0,
    'adult': 1,
    'adventure': 2,
    'animation': 3,
    'biography': 4,
    'comedy': 5,
    'crime': 6,
    'documentary': 7,
    'drama': 8,
    'family': 9,
    'fantasy': 10,
    'game-show': 11,
    'history': 12,
    'horror': 13,
    'music': 14,
    'musical': 15,
    'mystery': 16,
    'news': 17,
    'reality-tv': 18,
    'romance': 19,
    'sci-fi': 20,
    'short': 21,
    'sport': 22,
    'talk-show': 23,
    'thriller': 24,
    'war': 25,
    'western': 26
}

# Ensure the Genre column is in the correct format (e.g., lowercase and stripped of spaces)
df['Genre'] = df['Genre'].str.lower().str.strip()
# Use .loc to modify the DataFrame
df.loc[:, 'Genre_Number'] = df['Genre'].map(genre_mapping)

In [9]:
df.head()

Name     Genre  \
0       Oscar et la dame rose (2009)      drama   
1                       Cupid (1997)   thriller   
2   Young, Wild and Wonderful (1980)      adult   
3              The Secret Sin (1915)      drama   
4             The Unrecovered (2007)      drama   

                                         Description  Genre_Number  
0   Listening in to a conversation between his do...           8.0  
1   A brother and sister with a past incestuous r...          24.0  
2   As the bus empties the students for their fie...           1.0  
3   To help their unemployed father make ends mee...           8.0  
4   The film's title refers not only to the un-re...           8.0

## __STEP 2__: _Preprocessing the Description column using Spacy_

In [13]:
print(type(df['Genre']))

<class 'pandas.core.series.Series'>


In [31]:
import spacy
nlp = spacy.load("en_core_web_sm")

def preprocess(text):
    #remove stop words and lematize the text
    doc = nlp(text)
    filtered_tokens = []
    for token in doc:
        if token.is_stop or token.is_punct:
            continue
        filtered_tokens.append(token.lemma_)
    return " ".join(filtered_tokens)

In [33]:
#df['Description'] = df['Description'].fillna("").astype(str)
df = df.drop(labels= "PreProcessedDescription", axis = 1)
df["PreProcessedDescription"] = df['Description'].apply(preprocess)

In [34]:
df.head()

Name     Genre  \
0       Oscar et la dame rose (2009)      drama   
1                       Cupid (1997)   thriller   
2   Young, Wild and Wonderful (1980)      adult   
3              The Secret Sin (1915)      drama   
4             The Unrecovered (2007)      drama   

                                         Description  Genre_Number  \
0   Listening in to a conversation between his do...           8.0   
1   A brother and sister with a past incestuous r...          24.0   
2   As the bus empties the students for their fie...           1.0   
3   To help their unemployed father make ends mee...           8.0   
4   The film's title refers not only to the un-re...           8.0   

                             PreProcessedDescription  
0    listen conversation doctor parent 10 year ol...  
1    brother sister past incestuous relationship ...  
2    bus empty student field trip Museum Natural ...  
3    help unemployed father end meet Edith twin s...  
4    film title refer un recover body ground zero...

In [ ]:
from sklearn.model_selection import train_test_split
df['Genre_Number'] = df['Genre_Number'].fillna(-1)

## __STEP 3__: _Train Test Split_

In [35]:
X_train, X_test, y_train, y_test = train_test_split(df['PreProcessedDescription'], df['Genre'],test_size=0.2,random_state=42,stratify= df['Genre_Number'])

In [24]:
df['Genre_Number'].isna().sum()

0

In [23]:
df = df.dropna(subset=['Genre_Number'])

In [36]:
print('Shape of Xtrain is:',X_train.shape)

Shape of Xtrain is: (16660,)


In [30]:
y_train.value_counts()

drama          4531
documentary    3447
comedy         2430
short          1613
horror          699
thriller        549
action          389
western         355
reality-tv      284
family          253
adventure       238
romance         224
sci-fi          200
adult           195
music           167
crime           150
animation       141
talk-show       124
sport           114
fantasy         106
mystery          93
musical          74
biography        66
game-show        64
history          60
news             51
war              43
Name: Genre, dtype: int64

### __STEP 4__: Training the model using Naive Baye's Approach

In [41]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import TfidfVectorizer

clf = Pipeline([
    ('Vectorizer_tfidf', TfidfVectorizer()),
    ('Multi NB', MultinomialNB())   
])

clf.fit(X_train,y_train)

y_pred = clf.predict(y_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

      action       0.00      0.00      0.00        97
       adult       0.00      0.00      0.00        49
   adventure       0.00      0.00      0.00        60
   animation       0.00      0.00      0.00        35
   biography       0.00      0.00      0.00        17
      comedy       0.91      1.00      0.95       608
       crime       0.00      0.00      0.00        37
 documentary       0.76      1.00      0.86       862
       drama       0.58      1.00      0.73      1133
      family       0.00      0.00      0.00        63
     fantasy       0.00      0.00      0.00        26
   game-show       0.00      0.00      0.00        16
     history       0.00      0.00      0.00        15
      horror       0.00      0.00      0.00       175
       music       0.00      0.00      0.00        42
     musical       0.00      0.00      0.00        19
     mystery       0.00      0.00      0.00        23
        news       0.00    

C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## __NOTE__: For most of the classes that have small Training data such as romance(224), crime(150), war(43) etc. have zero F1_score.
## __WHY?__
### Because the data is highly imbalance, Classes with high training data like Drama, Documentry, Shorts dominates and had high F1_score i.e 0.73, 0.86, 1.00 respectively


In [43]:
y_train.value_counts()


drama          4531
documentary    3447
comedy         2430
short          1613
horror          699
thriller        549
action          389
western         355
reality-tv      284
family          253
adventure       238
romance         224
sci-fi          200
adult           195
music           167
crime           150
animation       141
talk-show       124
sport           114
fantasy         106
mystery          93
musical          74
biography        66
game-show        64
history          60
news             51
war              43
Name: Genre, dtype: int64

### Due to imbalance Dataset, I have decided to remove all entries that have value_counts() less than 250.
### Ideally the value_count() for all classes should be same. But this issue would be tackle using __'Handling imbalance class'__
__1ST STEP:__ Remove all classes with value count less than 250

In [51]:
df_filtered = df[~(df['Genre'].isin(['adventure', 'romance', 'sci-fi', 'adult', 'music', 'crime', 
                                     'animation', 'talk-show', 'sport', 'fantasy', 'mystery', 
                                     'musical', 'biography', 'game-show', 'history', 'news', 'war']))]


In [52]:
df_filtered['Genre'].value_counts()

drama          5664
documentary    4309
comedy         3038
short          2016
horror          874
thriller        686
action          486
western         444
reality-tv      355
family          316
Name: Genre, dtype: int64

__2nd STEP:__ Training the model on filtered data.

In [53]:
X_train, X_test, y_train, y_test = train_test_split(df_filtered['PreProcessedDescription'], df_filtered['Genre'],test_size=0.2,random_state=42,stratify= df_filtered['Genre_Number'])

In [54]:
clf = Pipeline([
    ('Vectorizer_tfidf', TfidfVectorizer()),
    ('Multi NB', MultinomialNB())   
])

clf.fit(X_train,y_train)

y_pred = clf.predict(y_test)

print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

      action       0.00      0.00      0.00        97
      comedy       1.00      1.00      1.00       608
 documentary       0.91      1.00      0.95       862
       drama       0.54      1.00      0.71      1133
      family       0.00      0.00      0.00        63
      horror       0.00      0.00      0.00       175
  reality-tv       0.00      0.00      0.00        71
       short       0.00      0.00      0.00       403
    thriller       0.00      0.00      0.00       137
     western       0.00      0.00      0.00        89

    accuracy                           0.72      3638
   macro avg       0.25      0.30      0.27      3638
weighted avg       0.55      0.72      0.61      3638



C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\hp\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [55]:
## Still the same problem ):, i need to find a solution to handle imbalance dataset. Apart from this my model is pretty DOPE.